### PR TITLE
Extend website

### PR DESCRIPTION
**Pull request by Naramsim**
This PR finalizes the skeleton for the Top10 website.

A live copy can be found at my repo: https://naramsim.github.io/Top10/

It also brings in a Github Action that builds and deploys the website on Github Pages. To enable Pages, @vanderaj  and @mackowski, on this very repository, go to _settings_ -> _pages_ and set everything as the screenshot below.

pages

---

@vanderaj and @mackowski, do you need a brief explanation of how everything works? Or will you understand it by looking at the files?

### CODE REVIEW

**Comment by Naramsim** [comment]
To test locally:

```sh
# Ensure Python 3
make install-python-requirements
make generate-site
# open http://localhost:8000/
```

**Comment by Naramsim** [comment]
I added as well a Github Action that builds and checks for dead links at every commit and on every pull request.

**Comment by mackowski** [comment]
@Naramsim looks awesome! Thank you :)
@vanderaj is a leader for this project so I will leave the final review for him.

**Comment by Naramsim** [comment]
Failing because of broken links:

- ('index.ja.md', 'master/2020/Data', 404)
- ('index.fr.md', 'master/2020/Data', 404)
- ('A04_2021-Insecure_Design.md', 'nist.gov/system/files/documents/2021/07/09/Developer%20Verification%20of%20Software.pdf', 404)
- ('index.zh_TW.md', 'master/2020/Data', 404)
- ('A09_2021-Security_Logging_and_Monitoring_Failures.ja.md', 'owasp.org/www-project-web-security-testing-guide/latest/4-Web_Application_Security_Testing/08-Testing_for_Error_Handling/01-Testing_for_Error_Code', 404)
- ('A09_2021-Security_Logging_and_Monitoring_Failures.zh_TW.md', 'owasp.org/www-project-web-security-testing-guide/latest/4-Web_Application_Security_Testing/08-Testing_for_Error_Handling/01-Testing_for_Error_Code', 404)
- ('A09_2021-Security_Logging_and_Monitoring_Failures.ja.md', 'owasp.org/www-project-web-security-testing-guide/latest/4-Web_Application_Security_Testing/08-Testing_for_Error_Handling/01-Testing_for_Error_Code', 404)

Can still be merged.

**Comment by Naramsim** [comment]
Fixed all the conflicts again.

Are there any plans @vanderaj to get this PR merged?

**Comment by mackowski** [comment]
hey @Naramsim I was on a trip without access to the internet. Yes I see that your changes are great. 
I am not in charge for this project @vanderaj @infosecdad @Neil-Smithline @sslHello - what do you think about this PR?

**Comment by sslHello** [comment]
Hi Alessandro (@Naramsim),
I am sorry that we have not replyed earlier.
Up to some weeks ago the upload of new gh-pages has been maintained by Andrew (@vanderaj).
As some translations were waiting to get published I've started to do it, too.

So I found your PR. Thank you very much, it looks very useful.
I'd have some questions:
* What happens if the gh-page gets broken by a pull?
* Do you think that it is possible to use it to generate the automatically generated pages as 'development' gh-page from it (how I do it locally)?  
* How do you get the languages sorted in the languages menue?

Thank you and cheers 
Torsten

**Comment by Naramsim** [comment]
Hi @sslHello,

1. If the build step passes, then it's impossible that the deploy step will generate and deploy a broken website. Anyway, if the GH deployment gets somehow broken, you can always revert it to a previous commit: https://github.com/OWASP/Top10/commits/gh-pages. In the end, GH is just serving static files from a specific git branch. 
2. I don't understand your sentence. Can you rephrase? 
3. It's a matter of ordering the dict in `mkdocs.yml`: https://github.com/Naramsim/Top10/blob/master/2021/mkdocs.yml#L87-L96 . Note that if running `mkdocs build` it works as expected, while if running `mkdocs serve` there is an unexpected behavior and the list isn't sorted correctly. But this happens only when developing locally.

**Comment by Naramsim** [comment]
Also note that the label with name _automatization_, could be renamed to the more common word _automation_.

**Comment by Naramsim** [comment]
Update on your 2nd point. Maybe I understand what you mean. 

Do you want a way to automatically generate the website, without generating it on your machine and deploying it manually? Well, this PR does exactly this. Whenever a PR will be merged in the `master`/`main` branch, a Github Action will build and deploy the website. Without any action needed by the maintainers. Please review the content on the `.github` folder: https://github.com/Naramsim/Top10/tree/master/.github/workflows

**Comment by Naramsim** [comment]
@sslHello, did you have time to review what I wrote?

**Comment by sslHello** [comment]
Hi @Naramsim,
thank you for your answeres. I am sorry I have been quite busy the last weeks.
* For my second question:<br>Is it possible to automatically deploy the content to a different branch and to compile a development site e.g. https://owasp.github.io/Top10-Dev/ and add some banners there like 'Development version and work in progress. "For the official OWASP Top 10 site visit 'https://owasp.org/Top10/', please." 
~ We'd like to publish new languages when they are more mature.
~ I am going to add 2 Mkdocs macros, that I've programmed in Python ('[https://github.com/sslHello/OSIB-Test](https://github.com/sslHello/OSIB-Test)'). Do you think it works? They read and write YAML files. The new, generated file extends the YAML structure (= Open Security Information Base, OSIB). It can be used by other versions, projects, documents and standards as universal references that fix even numbering changes between versions (if linked once in the structure).
* For my third question: Yes, thanks. As far as I see, I've experienced the unexpected behavior with `mkdocs serve` and `mkdocs gh-deploy`. So the list isn't sorted correctly at https://owasp.org/Top10/.

Thank you!
Cheers Torsten

**Comment by Naramsim** [comment]
One thing that can be made to finally include this PR in the source code is disabling the CI, so you won't have any automated deployments but we will have all the other changes.

**Comment by Naramsim** [comment]
Hi, @sslHello I came back here to see how the repo evolved. This PR can still be merged

**Comment by vanderaj** [comment]
Narasim, Can you please resolve the conflicts, and then step me through the GitHub actions because we've been attempted to be hacked by them in the past, so I am way of the way they work. We are starting work on the 2025 version, so your changes are very welcome, but I just need to see a version that will merge successfully, and then we should meet and you can take me through the way it works.

**Comment by Naramsim** [comment]
@vanderaj I fixed the conflicts. I'm still working on this but the `2021` folder is buldable. Please check on your machine if you can `mkdocs serve`, and please start with a clean venv and install the fixed dependencies found in the `2021/` folder.

**Comment by Naramsim** [comment]
@mackowski @sslHello Hi, I tried contacting @vanderaj in various (here, email, slack) ways but never ever received a reply from him. This is a very poor way of dealing with contributors willing to help you, especially considering that the Owasp foundation is based on open-source and should lead the way in terms of collaboration.

**Comment by mackowski** [comment]
@jmanico hey Jim can you help here?

**Comment by Naramsim** [comment]
Friendly ping to @vanderaj @mackowski
